### ISSUE TITLE
Bulk insert operations might be block thread When a node is disconnected from cluster

### ISSUE THREAD

**Issue by kkewwei** [elastic/elasticsearch#62676]
Elasticsearch version : 7.9.1
JVM version : 1.8.0_45
OS version : GNU/Linux
**Description of the problem including expected versus actual behavior:**
We use bulk interface of BulkProcessor to write data to the cluster, Considering the large amount of data, we set index.number_of_replicas to be 0, and shards of the index are assigned to every nodes of the cluster. During the writing process, a node goes offline,  then we find that the write threads are blocked.
we dump the thread stack:
1. Flush Thread
```
"elasticsearch[_client_][generic][T#1]" #82 daemon prio=5 os_prio=0 tid=0x00007f1725c02800 nid=0x214d waiting on condition [0x00007f17078fb000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00000000c1ff8440> (a java.util.concurrent.CountDownLatch$Sync)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
        at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
        at org.elasticsearch.action.bulk.BulkRequestHandler.execute(BulkRequestHandler.java:78)
        at org.elasticsearch.action.bulk.BulkProcessor.execute(BulkProcessor.java:454)
        at org.elasticsearch.action.bulk.BulkProcessor.execute(BulkProcessor.java:463)
        at org.elasticsearch.action.bulk.BulkProcessor.access$400(BulkProcessor.java:54)
        at org.elasticsearch.action.bulk.BulkProcessor$Flush.run(BulkProcessor.java:503)
```
2. Bulk Thread
```
"Source: kafkaReader -> Sink: eswriter (1/1)" #62 prio=5 os_prio=0 tid=0x00007f1724550800 nid=0x2059 waiting on condition [0x00007f170fdfe000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00000000c12dcf30> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:870)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1199)
        at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:209)
        at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:285)
        at org.elasticsearch.action.bulk.BulkProcessor.internalAdd(BulkProcessor.java:379)
        at org.elasticsearch.action.bulk.BulkProcessor.add(BulkProcessor.java:361)
        at org.elasticsearch.action.bulk.BulkProcessor.add(BulkProcessor.java:347)
```
we can see that two threads are waiting to get Lock. Bulk thread is block by ReentrantLock obtained by Flush thread, Flush thread is blocked by CountDownLatch.
 Why Flush thread can't get the Lock? I find that Flush thread send bulk request to a node ,but the node goes offline from cluster, and it doesn't response to Flush thread, also there has no timeout in bulk connection: https://github.com/elastic/elasticsearch/blob/b857768bb5bc1561c5c599897a9adb12048ee375/server/src/main/java/org/elasticsearch/action/bulk/BulkAction.java#L36
So the Flush thread will be blocked forever unless I restart the write process.

**Temporary plan**
I Use reflection to add timeout to BulkAction. If we could add timeout to BulkAction by settings not just Ignoring it.

**Comment by DaveCTurner** [comment]
This doesn't look to be the complete picture @kkewwei and certainly not enough for us to reproduce it. Requests always eventually receive a response, even if it's an error indicating that a connection was dropped, so the client likely wasn't blocked forever. How long did you wait?

Which node was dropped from the cluster? In particular, was it the coordinating node that this client was using, or was it a different node? Did it join the cluster again? Can you share the logs from the cluster?

Have you configured your [TCP retransmission timeout](https://www.elastic.co/guide/en/elasticsearch/reference/current/system-config-tcpretries.html) appropriately? The defaults are quite slow to detect a connection drop on Linux.

Are you using the deprecated transport client here?

**Comment by kkewwei** [comment]
Dropped node a data node. Because of hardware malfunction, the dropped node doesn't rejoin the cluster. net.ipv4.tcp_retries2 is default value(15). The write process is blocks more than two days before I restart it, 
the logs from master is as follow:
```
[2020-09-18T17:57:22,309][INFO ][o.e.c.s.ClusterApplierService] [maser] removed {{node1}{XM906CD1RjGPaRju_Wti1w}{MfFAX4EzSWGxNuIIxRLElQ}{10.xx.xx.22}{10.xx.xx22:9300}{dilrt}{ml.machine_memory=135299469312, rack=gha, ml.max_open_jobs=20, xpack.installed=true, transform.node=true}}, term: 1, version: 7593, reason: ApplyCommitRequest{term=1, version=7593, sourceNode={master1}{_SfHFfRpTimHlhUFJWrMgQ}{1B8FnKs8RnqPvCmqaRgoiA}{10.xx.xx.242}{10.xx.xx.242:9300}{ilmr}{ml.machine_memory=16644907008, rack=gha, ml.max_open_jobs=20, xpack.installed=true, transform.node=false}}
```
I don'y see any other wrong.
According to the phenomenon, it seems that:client sends bulk request to the data node, and data node has received the request correctly, then it processes the bulk request which costs much time, During the process, the data node goes offline.

**Comment by DaveCTurner** [comment]
Did the client send the bulk request _directly_ to the faulty data node, or did it send it to a different coordinating node?

When you say that the data node was "offline" do you mean it was still running and trying unsuccessfully to re-join the cluster for two whole days?

**Comment by kkewwei** [comment]
We don't set  coordinating node separately, any node can be coordinating node. Data node was "offline" mean the machine went down due to hardware problems, and we can't ping it.

Because of going offline, the data node can't response exception to the client, the netty channel in client might be disconnected from the data node after a few minutes, but the bulk request from the client will be hung forever.

**Comment by kkewwei** [comment]
it seem that client doesn't  take into account ongoing requests when closing the channel.

**Comment by DaveCTurner** [comment]
> It seems that client doesn't take into account ongoing requests when closing the channel.

That is very surprising. It should be returning a `NodeDisconnectedException` to all outstanding requests on a channel with the code here:

https://github.com/elastic/elasticsearch/blob/52ad5842a950332201efe852472c4f1b4d40a6ff/server/src/main/java/org/elasticsearch/transport/TransportService.java#L1034-L1035

How have you determined that this isn't happening in your case?

**Comment by kkewwei** [comment]
I find that the case can be recurrent. I try to analyze it and tell more details later.

**Comment by kkewwei** [comment]
I see the code: https://github.com/elastic/elasticsearch/blob/52ad5842a950332201efe852472c4f1b4d40a6ff/server/src/main/java/org/elasticsearch/transport/TransportService.java#L1034-L1035

For bulk request, it only comes here:
https://github.com/elastic/elasticsearch/blob/52ad5842a950332201efe852472c4f1b4d40a6ff/server/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java#L288
ES just pick another node out to resend the bulk request, and not interrupt the lock.

**Comment by DaveCTurner** [comment]
Sure, but that only tries once per node, and each attempt will receive some kind of response (even if just a `NodeDisconnectedException`) so this should still eventually complete the request.

**Comment by kkewwei** [comment]
the request can be eventually completed, but we don't unlock the ReentrantLock and CountDownLatch by interrupt, it can't be released

**Comment by DaveCTurner** [comment]
I'm still not following. When the request completes, the latch is unlocked:

https://github.com/elastic/elasticsearch/blob/ecd033bea62489e70c60763cfc117d0b96534fa0/server/src/main/java/org/elasticsearch/action/bulk/BulkRequestHandler.java#L74

(possibly after some more retries, according to your `BackoffPolicy`, but still only finitely many of them, so in any case it should return eventually)

**Comment by kkewwei** [comment]
I see the code:
https://github.com/elastic/elasticsearch/blob/ecd033bea62489e70c60763cfc117d0b96534fa0/server/src/main/java/org/elasticsearch/action/bulk/BulkRequestHandler.java#L74
will be executed after in `onResponse()` or after in `onFailure()`

`onResponse()` will be executed after the client receives the BulkResponse. `onFailure()` will be executed after failing to send the bulk request. The situation we face is that the bulk request has been sent successfully, `onFailure()` will not be executed, but as the data node went offline, the client cannot receive the response, so `onResponse()` will not be executed.
So `latch.countDown(); `will never be executed.

**Comment by DaveCTurner** [comment]
No, `onFailure()` is also executed when the request is sent successfully and then fails exceptionally (e.g. with a `NodeDisconnectedException`). At least that's how `client::bulk` behaves -- if you're using a different `consumer` with your `BulkProcessor` then you must ensure it behaves the same.

**Comment by kkewwei** [comment]
you are right, in `RetryListener. onFailure()` onFailure() may be executed, but on the one basic that:
1. there will be no available node to retry.
2. executing `callback.doWithNode(getNode(i), this);` and  throws exception
3. the exception is not `ConnectTransportException`
https://github.com/elastic/elasticsearch/blob/52ad5842a950332201efe852472c4f1b4d40a6ff/server/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java#L288

Once `callback.doWithNode(getNode(i), this);` is executed successfully, `onFailure()`will not be executed. This is the scene I described.

**Comment by DaveCTurner** [comment]
We already covered this in https://github.com/elastic/elasticsearch/issues/62676#issuecomment-696730906, but in more detail `callback.doWithNode(getNode(i), this)` passes `this` (i.e. a `RetryListener`) as its listener, which means that it will call `RetryListener#onFailure` again if the retry also fails (whether immediately or after successfully sending the request). It keeps doing this until `i >= nodes.size()` at which point it gives up and notifies the outer listener.

**Comment by kkewwei** [comment]
once returning a NodeDisconnectedException, the code will come to here: 
https://github.com/elastic/elasticsearch/blob/52ad5842a950332201efe852472c4f1b4d40a6ff/server/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java#L288

yes, you are right,  in `RetryListener` onFailure() may be executed, but on the one of the basic:
1. there will be no available node to retry.
2. executing `callback.doWithNode(getNode(i), this);` and  throws exception
3. the exception is not `ConnectTransportException`
https://github.com/elastic/elasticsearch/blob/52ad5842a950332201efe852472c4f1b4d40a6ff/server/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java#L288

Once `callback.doWithNode(getNode(i), this);` is executed successfully, `onFailure()`will not be executed. This is the scene I described.

(It's the comment I deleted, I wanted to give a more professional answer)

**Comment by kkewwei** [comment]
> It keeps doing this until i >= nodes.size() at which point it gives up and notifies the outer listener.
I Still have doubt, only if all nodes fail to send bulk request, the `listener.onFailure` can be executed, if `callback.doWithNode(getNode(i), this)` is successful at a node, it will not execute `listener.onFailure(inner);`

**Comment by DaveCTurner** [comment]
> if `callback.doWithNode(getNode(i), this)` is successful on a node, it will not execute `listener.onFailure(inner);` 

Yes that is right, if it's successful then it executes the `onResponse()` method of the listener instead. In either case the latch is released.

**Comment by kkewwei** [comment]
yes, you are right, code is ok.  I will close the issue, thank you for your support. Maybe there are other reason, I will go on try to find it.

**Comment by kkewwei** [comment]
@DaveCTurner, Recently it occurred again, the reason is that, when the client receives the bulkResponse, the worker thread crashes because of OutOfMemoryError, so the bulkResponse is dropped, as the result, the write thread is blocked in the code `latch.countDown(); `. we use Flink sink to write to es, it seems that Flink catch the exception, and the jvm works well.

**Comment by DaveCTurner** [comment]
This is a bug in your client: tou [should not catch subclasses of `java.lang.Error`](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/lang/Error.html). There's no reasonable way to proceed after an OOME except to shut the whole JVM down.
